### ISSUE TITLE
constraining number of media based on its type 

### ISSUE THREAD

**Issue by keighrim** [clamsproject/mmif#41]
In the prototype, we had `Audio`, `Image`, `Text`, and `Video` types. So the `media` field in the MMIF can hold multiple `medium` objects of different types. However, it doesn't sound very right that we have two or more `Video` or `Audio` inputs to a single MMIF file. However, as we discussed, if a tool can generate a new medium, that it's easy to imagine that a single MMIF file have multiple `Text` sources generated from different X-to-text tools (ASR, OCR) in the `media` list. Code-wise this must not be a big issue to pose some constrains based on the type, but we might want to add some justification and/or theory to the specification on this issue.

**Comment by marcverhagen** [comment]
That's an interesting point. Do we know enough about this to add constraints like that? Some videos have multiple audio tracks and I could come up with scenarios where we want all of them in MMIF. It does appear reasonable now that having multiple videos in one MMIF file will not happen, but is that always going to be the case and what exactly do we gain by restricting the number of videos to one? The one thing I can think of is that any video processing tool would not have to guess which of many videos to process.

We do seem to need a better way to know what exactly in the MMIF document to process since we have views and media and the LAPPS way for picking a view probably does not translate to CLAMS.

**Comment by keighrim** [comment]
Having spent some time on adding media and anchors to the vocabulary, I now think we can use the good old term `document` we used in LAPPS instead of `media`/`medium`, as defined as [this](http://vocab.lappsgrid.org/AudioDocument.html) or [that](http://vocab.lappsgrid.org/TextDocument.html). 

--- 
Anyway, going back to the topic here, first thing we need to figure out is how we tell apps to pick *correct* target document (medium) to process via the pipeline engine (galaxy at the moment?). Solving that should also solve the issue whether we want to allow multiple videos and/or audios (or any non-linguistic sources) in the media list.

**Comment by keighrim** [comment]
After today's discussion that we couldn't find a good justification for restricting numbers of video media only, we decided not to restrict the number of media objects in the list based on their types.

**Comment by keighrim** [comment]
With this change, I think now we need 
1. a way to *mark* a medium with its origin (e.g. passed as a primary input file? generated by some CLAMS app X?) that can be used in retrieving a specific medium. 
1. a default behavior of the retrieval API to get a medium when there's more than one media share the same type and origin. For reference, when retrieving a view the API returns the *latest* view with the type of interest.

**Comment by marcverhagen** [comment]
For marking the medium we already have something of a proposal in section 1.5 of https://mmif.clams.ai/0.1.0/:

```json
{
  "id": "m2",
  "type": "text",
  "text": {
    "@value": "yelp",
    "@language": "en"
  },
  "metadata": { 
    "source": "v1:bb1",
    "tool": "http://tools.clams.io/tesseract/1.2.1" 
  }
}
```

We may need to change the exact metadata though.

**Comment by keighrim** [comment]
a85bab83d2a01438b090b6f5bca6232aa57b17f3 implements the default behavior of getting a file location by a media type as returning the first of those match the type. It also adds some helper function to retrieve media based on `source` or `app`. Finally,  we decided to use the term `app`, instead of `tool` or `service`, I did some refactoring regarding that.

**Comment by keighrim** [comment]
closed via #74
